capellambse/_diagram_cache.py

```diff
@@ -0,0 +1,372 @@
+# SPDX-FileCopyrightText: Copyright DB Netz AG and the capellambse contributors
+# SPDX-License-Identifier: Apache-2.0
+"""Module to populate the diagram cache with native Capella diagrams."""
+
+from __future__ import annotations
+
+import datetime
+import json
+import logging
+import os
+import re
+import shlex
+import shutil
+import subprocess
+import tempfile
+import textwrap
+import typing as t
+from pathlib import Path
+
+from lxml import builder
+from lxml import etree as ET
+
+from capellambse.model.diagram import Diagram
+
+E = builder.ElementMaker()
+LOGGER = logging.getLogger(__name__)
+VALID_IMG_FMTS = (
+    "bmp",
+    "gif",
+    "jpg",
+    "png",
+    "svg",
+)
```

Comment:
OpenAI says:

A frozenset is immutable, meaning its elements cannot be modified after it is created. A tuple is also immutable, so the main advantage of a frozenset over a tuple is that it can be used as a key in a dictionary. Because frozensets are hashable, they can be used as elements of other sets or as members of a dictionary's keys, while a tuple cannot.

There is generally no significant difference in the performance of a frozenset and a tuple in Python. Both are immutable, so they are generally efficient for use cases where immutability is needed. The main difference in terms of performance is that because frozensets are hashable, they can be used as keys in dictionaries, which can lead to better performance when searching for an element in a dictionary with many items, but if the main usage of tuple or frozenset is for storing a small number of items and iterating them frozenset and tuple will have similar performance characteristics.

In general, when working with small numbers of items, the difference in performance between using a frozenset and a tuple is likely to be negligible, and the choice between the two should be based on their functional differences, rather than their performance characteristics.

Is it a good or a bad cop/ bot?

Reply:
see 0f80399

Reply:
That's a very long-winded way of saying "everything is fast for small enough n". :)

> the main advantage of a frozenset over a tuple is that it can be used as a key in a dictionary

Both tuples and frozensets can be `hash()`-ed, and so both can be used as dict keys; it's a perfectly valid use case to map e.g. a 2-tuple of strings to some callable. (What it probably meant is that _every_ possible frozenset can be used as key, while a tuple containing an unhashable object cannot – on the other hand, sets and frozensets cannot even contain unhashable objects in the first place.)

> There is generally no significant difference in the performance of a frozenset and a tuple in Python.

Unless we're talking specifically about very small sets and tuples (I don't know the exact question you asked), this is simply not true.

> Is it a good or a bad bot?

It's quite repetitive, and it only really made two points. However, these points are mostly valid, with the caveats above. I'd award a 6/10.

---

Personally I think it's a good idea to always use sets/frozensets for things that are meant to be used in containment checks. While I agree that the performance aspect taken by itself is negligible here (2µs more or less isn't gonna matter if the whole operation takes several minutes), I also recognize at least one other important point: communicating intent, both to other collaborators and your future self. If you see a frozenset here, you're going to immediately know "oh so these are the formats we support, we're going to check against this at some point later", while a tuple - at least to me - is more something to iterate over ("so are we going to always export all of these formats?"). Clearly communicated intentions make code generally easier to read and understand. (Which is essentially the same reason why we give names to variables, now that I think about it.)